README.md
# COMPARING SNEAKER API


<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/9/99/Unofficial_JavaScript_logo_2.svg/1024px-Unofficial_JavaScript_logo_2.svg.png" alt="drawing" width="30" />
<img src="https://scand.com/wp-content/uploads/2019/10/logo-node.png" alt="drawing" width="50" style="margin"/>

A sneaker API to get the best prices, services from a sneaker webshop.

## Requirements

You can use this api to compare sneaker prices on different webshops.

Use the package manager [npm](https://www.npmjs.com/) to install all the requirements.
-   Packages
    - [Puppeteer](https://www.npmjs.com/package/puppeteer)
    - [Express](https://www.npmjs.com/package/express)
    - [Cheerio](https://www.npmjs.com/package/cheerio)
    - [supertest](https://www.npmjs.com/package/supertest)
    - [request](https://www.npmjs.com/package/request)
    - [body-parser](https://www.npmjs.com/package/body-parser)
-   Testing
    - [Jest](https://www.npmjs.com/package/jest)


## Build containers
Open terminal on root of the document
```bash
cd api
docker compose up
```
open localhost:3000

## Test api
Open terminal on api and run
```bash
cd api
npm test 
```

## GET Endpoints
http://localhost:3000/seeds to run the seeders <br>
http://localhost:3000/show to show all the sneakers <br>
http://localhost:3000/torfs to start scraping torfs website. <br>
http://localhost:3000/brand/:brandname to show the brand by name<br>
http://localhost:3000/brandbyid/:uuid to show the brand by uuid<br>
http://localhost:3000/sneakers/:brand/:sort*?/ to show the sneaker by brand and sort it by name <br>


## POST Endpoints
http://localhost:3000/brand to create a brand. <br>

## PUT Endpoints
http://localhost:3000/updatebrand to update a brand. <br>

## DELETE Endpoints
http://localhost:3000/brand/:uuid to delete a brand. <br>

## Contributing
Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.<br>
contributing.md <br>
Please make sure to update tests as appropriate.

## License
[MIT](/LICENSE)

## Authors
This project was created by Andres Vergauwen, student Multimedia and Creative Technologies.

CHANGELOG.md
# V0.1.0

## Core
```
[GET] /  ➡ html file of api endpoint 
[GET] /torfs ➡ scraping endpoint
[POST] /torfs ➡ scraping endpoint
[POST] /brand ➡  create brand endpoint
[GET] /brandbyid/:uuid ➡ show brands by id endpoint
[GET] /brand/:brandname ➡ show brands by name endpoint
[PUT] /updatebrand ➡ update brand by uuid endpoint
[DELETE] /brand/:uuid ➡ delete brand by uuid endpoint
[GET] /seeds/ ➡  create seeders for brand and sneakers endpoint

```

# V0.2.0

## Core
```
[POST] /sneaker ➡ create a new sneaker endpoint
[GET] /sneakers/:brand/:sort ➡ show sneakers by brand and sorting endpoint
[DELETE] /sneaker/:uuid ➡ delete sneaker by uuid endpoint
[GET] /show ➡ show all sneakers endpoint
```

api/src/__tests__/unit.test.js
const Helpers = require('api/src/utils/helpers.js');
const Database = require('api/src/utils/database.js');
const sneakerObj = {
  uuid: Helpers.generateUUID(),
  product_brand: "Nike",
  product_name: "Air Force 1 '07",
  product_price: "€ 99,00",
  product_sale_price: "€ 89,00",
  product_sale: true,
  product_description: "Nike Air Force 1 '07",
  product_image: "https://www.snipes.be/dw/image/v2/BDCB_PRD/on/demandware.static/-/Sites-snse-master-eu/default/dwd508e7bc/1761737_P.jpg?sw=300&sh=300&sm=fit&sfrm=png",
  product_available: JSON.stringify([
    "36.5",
    "37.5",
    "38.5",
    "39.5",
    "40.5",
    "41.5",
  ]),
  product_colors: JSON.stringify(["White", "Black", "Red"]),
  product_url: "https://www.snipes.be/nl/p/nike-air_force_1_shadow-white%2Fwhite%2Fwhite-00013801761737.html",
  product_shipping_info: "8.7",
  brand_uuid: "6fed38e0-4d1d-11eb-9764-7b26be27a53d",
};
const brandObj = {
  uuid: Helpers.generateUUID(),
  brand_name: "zalando",
  brand_reviews: "6.7",
  brand_logo: "https://mosaic02.ztat.net/nvg/z-header-fragment/zalando-logo/logo_default.svg",
  brand_url: "https://www.zalando.be/",
};
describe('Check if BRAND object is correct', () => {
    test('check if brand_uuid is a uuid', async (done) => {
        try {
          const response = await Helpers.checkIfUuid(brandObj.uuid);
          expect(response).toBeTruthy()
          done()
        } catch (error) {
          console.log("❌ ERROR: ", error);
        }
      }),
      test('check if brand_name is a string', async (done) => {
        try {
          const response = await Helpers.checkIfString(brandObj.brand_name);
          expect(response).toBeTruthy()
          done()
        } catch (error) {
          console.log("❌ ERROR: ", error);
        }
      }),
      test('check if brand_logo is a valid url and image', async (done) => {
        try {
          const response = await Helpers.checkIfString(brandObj.brand_logo);
          expect(response).toBeTruthy()
          done()
        } catch (error) {
          console.log("❌ ERROR: ", error);
        }
      }),
      test('check if brand_url is a valid url', async (done) => {
        try {
          const response = await Helpers.checkIfString(brandObj.brand_url);
          expect(response).toBeTruthy()
          done()
        } catch (error) {
          console.log("❌ ERROR: ", error);
        }
      })
  }),
  describe('Check if SNEAKER object is correct', () => {
    test('check if product_brand is a string', async (done) => {
        try {
          const response = await Helpers.checkIfString(sneakerObj.product_brand);
          expect(response).toBeTruthy()
          done()
        } catch (error) {
          console.log("❌ ERROR: ", error);
        }
      }),
      test('check if product_name is a string', async (done) => {
        try {
          const response = await Helpers.checkIfString(sneakerObj.product_name);
          expect(response).toBeTruthy()
          done()
        } catch (error) {
          console.log("❌ ERROR: ", error);
        }
      }),
      test('check if product_price has a euro in front of number', async (done) => {
        try {
          const response = await Helpers.checkIfPriceHasEuro(sneakerObj.product_price);
          expect(response).toBeTruthy()
          done()
        } catch (error) {
          console.log("❌ ERROR: ", error);
        }
      }),
      test('check if product_sale_price has a euro in front of number', async (done) => {
        try {
          const response = await Helpers.checkIfPriceHasEuro(sneakerObj.product_sale_price);
          expect(response).toBeTruthy()
          done()
        } catch (error) {
          console.log("❌ ERROR: ", error);
        }
      }),
      test('check if product_sale is a boolean', async (done) => {
        try {
          const response = await Helpers.checkIfBoolean(sneakerObj.product_sale);
          expect(response).toBeTruthy()
          done()
        } catch (error) {
          console.log("❌ ERROR: ", error);
        }
      }),
      test('check if product_image is a image and url', async (done) => {
        try {
          const response = await Helpers.checkIfUrlImage(sneakerObj.product_image);
          expect(response).toBeTruthy()
          done()
        } catch (error) {
          console.log("❌ ERROR: ", error);
        }
      });
    test('check if product_available is an array', async (done) => {
      try {
        const response = await Helpers.checkIfArray(sneakerObj.product_available);
        expect(response).toBeTruthy()
        done()
      } catch (error) {
        console.log("❌ ERROR: ", error);
      }
    });
    test('check if product_colors is an array', async (done) => {
      try {
        const response = await Helpers.checkIfArray(sneakerObj.product_colors);
        expect(response).toBeTruthy()
        done()
      } catch (error) {
        console.log("❌ ERROR: ", error);
      }
    });
    test('check if brand_uuid is an uuid', async (done) => {
      try {
        const response = await Helpers.checkIfUuid(sneakerObj.brand_uuid);
        expect(response).toBeTruthy()
        done()
      } catch (error) {
        console.log("❌ ERROR: ", error);
      }
    });
    test('check if product_url is a url', async (done) => {
      try {
        const response = await Helpers.checkIfUrl(sneakerObj.product_url);
        expect(response).toBeTruthy()
        done()
      } catch (error) {
        console.log("❌ ERROR: ", error);
      }
    });
  })
 

api/src/__tests__/helper.test.js

const Helpers = require('api/src/utils/helpers.js');

describe('stringlength tester', () => {
  test('if there is a string', () => {
    expect (Helpers.checkTitleLength()).toBeFalsy()
    expect (Helpers.checkTitleLength(102)).toBeFalsy()
    expect (Helpers.checkTitleLength([])).toBeFalsy()
  })

  test('if string length is good', () => {
    expect(Helpers.checkTitleLength("Hello world").length).toBeLessThan(51);
    expect(Helpers.checkTitleLength("Hello world, how is it going, this is a ti").length).toBeLessThan(51);
  })
  test('If string starts with a capital', () => {
    expect(Helpers.checkTitleLength("hello wolrd")).toBeFalsy();
  }) 
})



api/src/__tests__/end_to_end.test.js
//===================================================================================================================================================================================================================================================================================================================
//                                                                                                                                                                                                                                                                                                                   
//    ###    ##     ##  ####    #####    #####   ####                                                                                                                                                                                                                                                              
//   ## ##   ####   ##  ##  ##  ##  ##   ##     ##                                                                                                                                                                                                                                                                 
//  ##   ##  ##  ## ##  ##  ##  #####    #####   ###                                                                                                                                                                                                                                                               
//  #######  ##    ###  ##  ##  ##  ##   ##        ##                                                                                                                                                                                                                                                              
//  ##   ##  ##     ##  ####    ##   ##  #####  ####                                                                                                                                                                                                                                                               
//                                                                                                                                                                                                                                                                                                                   
//===================================================================================================================================================================================================================================================================================================================
const supertest = require('supertest')
const Helpers = require('api/src/utils/helpers.js');
const app = require('../server.js')
const request = supertest(app)
let uuid = Helpers.generateUUID();
/**
 * ====== FLOW ======
 * create brand
 * try to create same brand 
 * show the created brand
 * create sneaker with brand relation
 * create other sneaker with brand relation
 * show the sneaker with the created brand
 * update the brand
 * delete de brand
 * try to get the sneaker => has to be deleted
 * check if brand is deleted
 */

let brandObj = {
    "uuid": uuid,
    "brand_name": "adidas",
    "brand_reviews": "9.7",
    "brand_logo": "https://www.adidas.be/glass/react/755aab5/assets/img/icon-adidas-logo.svg",
    "brand_url": "https://www.adidas.com/"
};
let sneakerObj = {
    uuid: Helpers.generateUUID(),
    product_brand: "Nike",
    product_name: "Air Force 1 '07",
    product_price: "€ 99,00",
    product_sale_price: "€ 89,00",
    product_sale: true,
    product_description: "Nike Air Force 1 '07",
    product_image: "https://www.snipes.be/dw/image/v2/BDCB_PRD/on/demandware.static/-/Sites-snse-master-eu/default/dwd508e7bc/1761737_P.jpg?sw=300&sh=300&sm=fit&sfrm=png",
    product_available: JSON.stringify([
        "36.5",
        "37.5",
        "38.5",
        "39.5",
        "40.5",
        "41.5",
    ]),
    product_colors: JSON.stringify(["White", "Black", "Red"]),
    product_url: "https://www.snipes.be/nl/p/nike-air_force_1_shadow-white%2Fwhite%2Fwhite-00013801761737.html",
    product_shipping_info: "8.7",
    brand_uuid: uuid
};
let sneakerObj1 = {
    uuid: Helpers.generateUUID(),
    product_brand: "Nike 2",
    product_name: "Air Force 2 '07",
    product_price: "€ 122,00",
    product_sale_price: "€ 99,00",
    product_sale: true,
    product_description: "Nike Air Force 2 '07",
    product_image: "https://www.snipes.be/dw/image/v2/BDCB_PRD/on/demandware.static/-/Sites-snse-master-eu/default/dwd508e7bc/1761737_P.jpg?sw=300&sh=300&sm=fit&sfrm=png",
    product_available: JSON.stringify([
        "34.5",
        "37.5",
        "38.5",
        "42.5",
        "40.5",
        "41.5",
    ]),
    product_colors: JSON.stringify(["White", "Black", "Green"]),
    product_url: "https://www.snipes.be/nl/p/nike-air_force_1_shadow-white%2Fwhite%2Fwhite-00013801761737.html",
    product_shipping_info: "5.7",
    brand_uuid: uuid
};

describe('Flow of brands and sneakers', () => {
    test('create a brand', async (done) => {
            try {
                const response = await request.post('/brand').send(brandObj)
                expect(response.status).toBe(201)
                done()
            } catch (error) {
                console.log("❌ ERROR: ", error);
            }
        }),
        test('check if brand is not created beacause of duplicate uuid', async (done) => {
            try {
                const response = await request.post('/brand').send(brandObj)
                expect(response.status).toBe(400)
                done()
            } catch (error) {
                console.log("❌ ERROR: ", error);
            }
        }),
        test('get the created brand?', async (done) => {
            try {
                const response = await request.get(`/brandbyid/${uuid}`)
                expect(response.status).toBe(200)
                done()
            } catch (error) {
                console.log("❌ ERROR: ", error);
            }
        }),
        test('create a sneaker', async (done) => { 
            try {
                const response = await request.post('/sneaker').send(sneakerObj)
                expect(response.status).toBe(201)
                done()
            } catch (error) {
                console.log("❌ ERROR: ", error);
            }
        }),
        test('create another sneaker', async (done) => { 
            try {
                const response = await request.post('/sneaker').send(sneakerObj1)
                expect(response.status).toBe(201)
                done()
            } catch (error) {
                console.log("❌ ERROR: ", error);
            }
        }),

        test('show sneakers by brand', async (done) => { 
            try {
                const response = await request.get(`/sneakers/${brandObj.brand_name}`);
                expect(response.status).toBe(200)
                done()
            } catch (error) {
                console.log("❌ ERROR: ", error);
            }
        }),
        test('update the brand', async (done) => {
            try {
                brandObj.brand_name = "bol";
                const response = await request.put(`/updatebrand`).send(brandObj)
                expect(response.status).toBe(204)
                done()
            } catch (error) {
                console.log("❌ ERROR: ", error);
            }
        }),
        test('delete the brand', async (done) => {
            try {
                const response = await request.delete(`/brand/${uuid}`)
                expect(response.status).toBe(204)
                done()
            } catch (error) {
                console.log("❌ ERROR: ", error);
            }
        }),
        test('check if sneakers are deleted', async (done) => {
            try {
                const response = await request.get(`/sneakers/${brandObj.brand_name}`)
                expect(response.status).toBe(400)
                done()
            } catch (error) {
                console.log("❌ ERROR: ", error);
            }
        }),
        test('check if brand does not exist anymore', async (done) => {
            try {
                const response = await request.get(`/brandbyid/${uuid}`)
                expect(response.status).toBe(400)
                done()
            } catch (error) {
                console.log("❌ ERROR: ", error);
            }
        })
})

api/src/index.js
//===================================================================================================================================================================================================================================================================================================================
//                                                                                                                                                                                                                                                                                                                   
//    ###    ##     ##  ####    #####    #####   ####                                                                                                                                                                                                                                                              
//   ## ##   ####   ##  ##  ##  ##  ##   ##     ##                                                                                                                                                                                                                                                                 
//  ##   ##  ##  ## ##  ##  ##  #####    #####   ###                                                                                                                                                                                                                                                               
//  #######  ##    ###  ##  ##  ##  ##   ##        ##                                                                                                                                                                                                                                                              
//  ##   ##  ##     ##  ####    ##   ##  #####  ####                                                                                                                                                                                                                                                               
//                                                                                                                                                                                                                                                                                                                   
//===================================================================================================================================================================================================================================================================================================================

const app = require("./server.js");
const port = 3000;

app.listen(port, () => {
    console.info("✅", `Listening to requests on http://localhost:${port}`);
  });

api/src/utils/helpers.js
const {
  v1: uuidv1
} = require('uuid');

const Helpers = {
  generateUUID: () => {
    const uuid = uuidv1();
    return uuid;
  },
  checkContentLength: (title, maxLength) => {
    if (title.length > maxLength) {
      return false
    }
    return true
  },
  checkIfString: (title) => {
    if (typeof title === "string") {
      return true
    }
    return false
  },
  checkIfPriceHasEuro: (price) => {
    if (price.startsWith("€")) {
      return true
    }
    return false
  },
  checkIfUrl: (url) => {
    let pattern = new RegExp('^(https?:\\/\\/)?' + // protocol
      '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|' + // domain name
      '((\\d{1,3}\\.){3}\\d{1,3}))' + // OR ip (v4) address
      '(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*' + // port and path
      '(\\?[;&a-z\\d%_.~+=-]*)?' + // query string
      '(\\#[-a-z\\d_]*)?$', 'i'); // fragment locator
    return !!pattern.test(url);
  },
  checkIfImage: (image) => {
    return image.match(/\.(jpeg|jpg|png|gif|svg)/g) != null
  },
  checkIfUrlImage: (url) => {
    let checkIfUrl = Helpers.checkIfUrl(url);
    let checkIfImage = Helpers.checkIfImage(url);
    return !!checkIfUrl && !!checkIfImage;
  },
  checkIfArray: (array) => {
    return (typeof JSON.parse(array) === "object");
  },
  checkIfBoolean: (boolean) => {
    return (typeof boolean === "boolean");
  },
  checkIfUuid: (uuid) => {
    let pattern = new RegExp(/^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/);
    return !!pattern.test(uuid);
  },

  checkSneakerObj: (sneakerObj) => {
    let checkUuid = Helpers.checkIfUuid(sneakerObj.uuid);
    let checkProductBrand = Helpers.checkIfString(sneakerObj.product_brand);
    let checkProductName = Helpers.checkIfString(sneakerObj.product_name);
    let checkProductPrice = Helpers.checkIfPriceHasEuro(sneakerObj.product_price);
    let checkProductSalePrice = Helpers.checkIfPriceHasEuro(sneakerObj.product_sale_price);
    let checkProductSale = Helpers.checkIfBoolean(sneakerObj.product_sale);
    let checkProductImage = Helpers.checkIfUrlImage(sneakerObj.product_image);
    let checkProductUrl = Helpers.checkIfUrl(sneakerObj.product_url);
    let checkProductAvailable= Helpers.checkIfArray(sneakerObj.product_available);
    let checkProductColors= Helpers.checkIfArray(sneakerObj.product_colors);
    let checkBrandUuid= Helpers.checkIfUuid(sneakerObj.brand_uuid);
    if(!checkUuid || !checkProductBrand || !checkProductName || !checkProductPrice||!checkProductSalePrice || !checkProductUrl || !checkProductSale || !checkProductImage || !checkProductAvailable|| !checkProductColors|| !checkBrandUuid){
      return false;
    }
    return true;
  },
  checkBrandObj: (brandObj) => {
    let checkUuid = Helpers.checkIfUuid(brandObj.uuid);
    let checkBrandName = Helpers.checkIfString(brandObj.brand_name);
    let checkBrandImage = Helpers.checkIfUrlImage(brandObj.brand_logo);
    let checkBrandUrl = Helpers.checkIfUrl(brandObj.brand_url);
    if(!checkUuid || !checkBrandName || !checkBrandImage || !checkBrandUrl){
      return false;
    }
    return true;
  },
}





module.exports = Helpers

api/src/utils/database.js
const Helpers = require("./helpers.js");
const pg = require("knex")({
  client: "pg",
  version: "9.6",
  searchPath: ["knex", "public"],
  connection: process.env.PG_CONNECTION_STRING ?
    process.env.PG_CONNECTION_STRING : "postgres://example:example@localhost:5432/sneakerdb",
});

const database = {
  /*--------- INITIALIZE TABLES --------*/
  initialiseTables: async () => {
    await pg.schema.hasTable("brands").then(async (exists) => {
      if (!exists) {
        await pg.schema
          .createTable("brands", (table) => {
            table.increments();
            table.uuid("uuid").notNullable().unique();
            table.string("brand_name");
            table.string("brand_reviews");
            table.text("brand_logo");
            table.text("brand_url");
            table.timestamps(true, true);
          })
          .then(async () => {
            console.log("🎉", "created brands table");
            database.brandSeeders();
          });
      }
    });
    await pg.schema.hasTable("sneakers").then(async (exists) => {
      if (!exists) {
        await pg.schema
          .createTable("sneakers", (table) => {
            table.increments();
            table.uuid("uuid").notNullable().unique();
            table.string("product_brand");
            table.string("product_name");
            table.string("product_price");
            table.string("product_sale_price");
            table.boolean("product_sale");
            table.text("product_description");
            table.text("product_image");
            table.text("product_available");
            table.text("product_colors");
            table.text("product_url");
            table.string("product_shipping_info");
            table
              .uuid("brand_uuid")
              .unsigned()
              .references("uuid")
              .inTable("brands")
              .onDelete("CASCADE")
              .onUpdate("CASCADE")
              .notNullable();
            table.timestamps(true, true);
          })
          .then(async () => {
            console.log("🎉", "created sneakers table");
            database.sneakerSeeders();
            process.exit();
          });
      }
    });
  },

  /*--------- SEEDERS --------*/
  sneakerSeeders: async () => {
    const sneakerObj = {
      uuid: Helpers.generateUUID(),
      product_brand: "Nike",
      product_name: "Air Force 1 '07",
      product_price: "€ 99,00",
      product_sale_price: "€ 89,00",
      product_sale: true,
      product_description: "Nike Air Force 1 '07",
      product_image: "https://www.snipes.be/dw/image/v2/BDCB_PRD/on/demandware.static/-/Sites-snse-master-eu/default/dwd508e7bc/1761737_P.jpg?sw=300&sh=300&sm=fit&sfrm=png",
      product_available: JSON.stringify([
        "36.5",
        "37.5",
        "38.5",
        "39.5",
        "40.5",
        "41.5",
      ]),
      product_colors: JSON.stringify(["White", "Black", "Red"]),
      product_url: "https://www.snipes.be/nl/p/nike-air_force_1_shadow-white%2Fwhite%2Fwhite-00013801761737.html",
      product_shipping_info: "8.7",
      brand_uuid: "6fed38e0-4d1d-11eb-9764-7b26be27a53d",
    };
    const sneakers = await pg
      .table("sneakers")
      .insert(sneakerObj)
      .then(async function () {
        console.log("✅", "Created new sneaker");
        return;
      })
      .catch((error) => {
        console.log("❌ ERROR: ", error.message);
      });
  },
  brandSeeders: async () => {
    const brandArrayObj = [{
        uuid: Helpers.generateUUID(),
        brand_name: "snipes",
        brand_reviews: "8.7",
        brand_logo: "https://www.snipes.nl/on/demandware.static/Sites-snse-NL-BE-Site/-/default/dwcc537b29/images/snipes_logo.svg",
        brand_url: "https://www.snipes.com/",
      },
      {
        uuid: Helpers.generateUUID(),
        brand_name: "sneakerdistrict",
        brand_reviews: "8.7",
        brand_logo: "https://www.sneakerdistrict.fr/assets/img/sneaker-district-weblogo@2x.png",
        brand_url: "https://www.sneakerdistrict.nl/",
      },
      {
        uuid: "6fed38e0-4d1d-11eb-9764-7b26be27a53d",
        brand_name: "torfs",
        brand_reviews: "8.7",
        brand_logo: "https://www.torfs.be/on/demandware.static/Sites-Torfs-Webshop-BE-Site/-/default/dw1f8272ff/images/logo_standard.svg",
        brand_url: "https://www.torfs.be/",
      },
    ];
    for (const brandObj of brandArrayObj) {
      const brands = await pg
        .table("brands")
        .insert(brandObj)
        .then(async function () {
          console.log("✅", "Created new brand");
        })
        .catch((error) => {
          console.log("❌ ERROR: ", error.message);
        });
    }
  },

  /*--------- CRUD  --------*/
  // Brand
  addBrand: async (brandObj) => {
    let result;
    let checkBrandObj = Helpers.checkBrandObj(brandObj)
    if (!checkBrandObj) {
      return false;
    }
    const brand = await pg
      .table("brands")
      .insert(brandObj)
      .then(function () {
        result = true;
      })
      .catch((error) => {
        result = false;
      });
    return result;
  },
  getBrandByName: async (brandName) => {
    let brand = await pg
      .from("brands")
      .where({
        brand_name: brandName
      });
    return brand;
  },
  getBrandById: async (brandId) => {
    try {
      const brand = await pg.select().from("brands").where({
        uuid: brandId
      });
      return brand;
    } catch (error) {
      console.log("❌ ERROR: ", error);
    }
  },
  updateBrandById: async (brandObj) => {
    try {
      let checkBrandObj = Helpers.checkBrandObj(brandObj)
      if (!checkBrandObj) {
        return false
      }
      const brand = await pg
        .table("brands")
        .where({
          uuid: brandObj.uuid
        })
        .update(brandObj);
      return true;
    } catch (error) {
      console.log("❌ ERROR: ", error.message);
    }
  },
  deleteBrandById: async (brandId) => {
    try {
      let checkBrandUuid = Helpers.checkIfUuid(brandId)
      if (!checkBrandUuid) {
        return false
      }
      const brand = await pg.table("brands").where({
        uuid: brandId
      }).del();
      return true;
    } catch (error) {
      console.log("❌ ERROR: ", error.message);
    }
  },
 
  // Sneakers
  addSneakers: async (sneakersArray) => {
    const sneakers = await pg
      .table("sneakers")
      .insert(sneakersArray)
      .then(function () {
        console.log(
          "✅",
          `Succesfully Created ${sneakersArray.length} sneakers`
        );
        return;
      })
      .catch((error) => {
        console.log("❌ ERROR: ", error.message);
      });
  },
  addSneaker: async (sneakerObj) => {
    let checkSneakerObj = Helpers.checkSneakerObj(sneakerObj)
    if (!checkSneakerObj) {
      return false
    }

    const sneakers = await pg
      .table("sneakers")
      .insert(sneakerObj)
      .then(function () {
        return true;
      })
      .catch((error) => {
        console.log("❌ ERROR: ", error);
      });
    return true;
  },
  getSneakersByBrand: async (brandName, sort) => {
    try {
      const sneakers = await pg
        .select([
          "brands.brand_name",
          "brands.brand_logo",
          "brands.brand_url",
          "brands.brand_reviews",
          "sneakers.product_brand",
          "product_name",
          "product_price",
          "product_sale_price",
          "product_sale",
          "product_description",
          "product_image",
          "product_available",
          "product_url",
          "product_shipping_info",
        ])
        .from("brands")
        .rightJoin("sneakers", "sneakers.brand_uuid", "brands.uuid")
        .where({
          brand_name: brandName.toLocaleLowerCase()
        })
        .orderBy("product_name", `${sort}`);
      return sneakers;
    } catch (error) {
      console.log("❌ ERROR: ", error.message);
    }
  },
  deleteSneaker: async (sneakerUuid) => {
    let checkSneakerUuid = Helpers.checkIfUuid(sneakerUuid)
    if (!checkSneakerUuid) {
      return false
    }
    const sneakers = await pg
      .table("sneakers")
      .where({
        uuid: sneakerUuid
      })
      .del()
      .then(async function () {
        console.log("✅", "Sneaker has been deleted");
        return true;
      })
      .catch((error) => {
        console.log("❌ ERROR: ", error.message);
      });
  },
  deleteSneakers: async () => {
    const sneakers = await pg
      .table("sneakers")
      .truncate()
      .then(async function () {
        console.log("✅", "Sneakers has been truncated");
        return;
      })
      .catch((error) => {
        console.log("❌ ERROR: ", error.message);
      });
  },
};
module.exports = database;